src/main/perl/Component.pm

```diff
@@ -226,8 +227,7 @@ sub Configure
 {
     my ($self, $config) = @_;
 
-    $self->error('Configure() method not implemented by component');
-    return;
+    return $self->_redirect("Configure", $config);
```

Comment:
How does this work? Previously the code was only called if the subclass didn't override the Configure() method. So how does a module author use this new facility?

Reply:
if no redirect is defined, it still reports an error as before. that behaviour has not changed.
i'm sure i fully understand what you want to know. you can use the new facility by setting the REDIRECT variable. i can clarify the pod below if needed

Reply:
just to give you an idea: before you had to do https://github.com/quattor/configuration-modules-core/blob/master/ncm-spma/src/main/perl/spma.pm; now you can do https://github.com/kwaegema/configuration-modules-core/blob/5fb11b7f293a920e6c63282684a20a804e66c472/ncm-ceph/src/main/perl/ceph.pm

Reply:
Thanks for the example, but that doesn't really help! It seems very unintuitive when reading the code that the parameters in a Readonly hash relate to a profile attribute that then select a file from a subdirectory which is not even named the same as the file. (Ceph/ v ceph). It's much less code but a significantly higher barrier to anyone approaching the code to figure out what's happening.

In particular using a hash to ultimately invoke a method is unintuitive and breaks the OO concept of inheritance. I think we should keep to using inheritance or other well-understood OO concepts. e.g. 

```
# This components use a sub-module to select the right code to run
sub Configure {
  my ($self, $config) = @_;
  my $submod = $config->getTree("/path/to/module");
  $submod = "Luminous" unless defined $submod;
  return $self->redirect($submod);  # redirect is a method in the super class.
```

feels more naturally OO for not significantly more code.

Reply:
this has been a while, but i don't agree with your view

in your example, the `Luminous` and `/path/to/module` are proper constants and as such should be Readonly anyway; so you end up with code like
```perl
Readonly $PATH => '/path/to/module';
Readonly $DEFAULT => 'Luminous';

sub Configure {
my ($self, $config) = @_;
  my $submod = $config->getTree($PATH);
  $submod = $DEFAULT unless defined $submod;
  return $self->redirect($submod);  # redirect is a method in the super class.
}
```

now, what is so OO about having to copy the `Configure` method each time? (and for completeness, the example is missing a bunch of error handling and with fixed `$PATH` doesn't allow aliasing, so the full code block to copy/paste will be even bigger)
your other remarks wrt "profile attribute" and naming, that is legacy quattor code. there are no standards, i would definitley prefer not haing to do this and just use the same for all, but for some reaon our base component classes have lower case names, and i can hardly use `packager` as an attribute name in ceph to point out a certain version/flavour of the software. 

maybe if we can settle on a proper attribute name and it's ok to have components start with uppercase, i can make the code do its magic based on single `$REDIRECT_DEFAULT` reaonly instead of a hashref?

Reply:
My issue is with this being completely a data definition:

```
Readonly our $REDIRECT => {
    name => 'release',
    default => 'Luminous',
};
```

If I am trying to follow this code I am completely stuck. I need to grep the code base for where  $REDIRECT is used and then understand the abstraction. Compare this to my example, I can see instantly (because I am familiar with the standard OO abstraction) that I need to check the definition of the `redirect` function in the super-class.